JIT compilation, first cut, squashed

hrj opened the pull request:
This PR is very similar to #440, but with cleaned up commits.

My original description below:

------

This is what I have been working on for the last few weeks!

Design wise, this is the first-cut, simple version. I have been testing/benchmarking/tweaking the implementation to see how far the design goes, so the implementation is quite stable.

The main purpose of the PR is to get some feedback from @jvilk, et al. Though, I believe it is stable enough to merge (after review).

## Design Overview

Currently, Doppio *interprets* JVM byte code one at a time. The opcodes are stored in an array and the interpreter walks the elements of the array sequentially. Each op-code is executed by invoking a corresponding function. The function is picked through a table indexed by the opcode.

One strategy for improving performance is to compile a sequence of opcodes into an equivalent JS function.

For example, consider this simple Java function:

```
public static int add(int a, int b, int c) {
  return a + b + c;
}
```

which after compilation produces this byte code:

```
public static int add(int, int, int);
  Code:
     0: iload_0
     1: iload_1
     2: iadd
     3: iload_2
     4: iadd
     5: ireturn
```

Instead of interpreting these opcodes one at a time, one could compile a JS function which does that:
```
function add() {
  iload_0();
  iload_1();
  iadd();
  iload_2();
  iadd();
  ireturn();
}
```

This avoids checking the opcodes of the method everytime. However, we can go further.  We can have stackless variants of the opcode implementations. For example, `stackless_iload_0` would return the loaded value rather than push it onto the stack. Similarly, `stackless_iadd` would accept two parameters and return the result, instead of using the stack. The stackless variant would look like:

```
function add() {
  const s0 = stackless_iload_0();
  const s1 = stackless_iload_1();
  const s2 = stackless_iadd(s0, s1);
  const s3 = stackless_iload_2();
  const s4 = stackless_iadd(s2, s3);
  stackless_ireturn(s4);
}
```

Further, the stackless functions could be inlined at the time of emitting, to give this form:

```
function add(frame, thread) {
  const s0 = frame.locals[0];
  const s1 = frame.locals[1];
  const s2 = s0 + s1;
  const s3 = frame.locals[2];
  const s4 = s2 + s3;
  thread.return(s4);
}
```

This form should improve performance since stack operations have been replaced with local variable assignments and the
Javascript VM can optimise this function fully.

Such a function can be emitted as a string and then compiled using JS `eval` or `new Function()`.

The above example is simplistic. There are some gotchas to consider:

*  If the function contains jumps then such a straightforward translation to JS is not possible.
*  If checks for yielding the thread are added, the continuation functions would create too many closures.

One solution is to create separate functions for each "trace", where a trace is defined as a contiguous block of opcodes that don't branch nor yield. A Java function might comprise of multiple such traces, and hence would have multiple JS functions associated with it.


## Tradeoffs

* Identifying traces, generating their JS functions and compiling those functions will increase the warm up time.
* Creating these JS functions would increase memory consumption.

To mitigate these factors, JITing can be restricted to frequently called functions. Currently a very simple heuristic is used: if the count of opcode executions in a method exceeds a threshold, then that method is targeted for JIT compilation.

## Implementation details

### Symbolic stack
To keep track of symbols needed for stackless execution, a symbol stack is maintained similar to the operand stack. The difference is that the symbol stack tracks symbols at JIT time, while the operand stack tracks actual values at run-time.

When an opcode needs to pop two values from the operand stack, the JIT will instead pop two symbols from the symbol stack. When the symbol stack underflows, new symbols are created and initialised from the operand stack.

Similarly, when an opcode needs to push a value to the operand stack, the JIT emitted code will assign the value to a symbol which is then pushed on the symbol stack. This symbol can then be consumed by later opcodes in the `trace`. At the end of the trace, all symbols remaining on the symbol stack are mapped to pushes on the operand stack.

As an example, consider a trace that contains the following opcodes: `IADD`, `IMUL`. The first opcode needs two symbols from the symbolic stack. Since the symbolic stack is initially empty, we create two symbols, s1 and s2, push them to the symbolic stack, and emit code to initialise them:

```
var s1 = frame.opStack.pop();
var s2 = frame.opStack.pop();
```

These symbols are then available to the the `IADD` instruction, for which the following code can be emitted:

```
var s3 = s1 + s2;
```

Here, `s3` is a new symbol that represents the pushed value from `IADD`. This symbol will now be pushed on the symbol stack. When translating the next opcode `IMUL`, we again need two symbols to be popped from the symbol stack. But the stack contains only one symbol `s3`. Hence, a new symbol `s4` is allocated and pushed onto the stack, and code to initialise is emitted:

```
var s4 = frame.opStack.pop();
```

followed by the code for `IMUL`:

```
var s5 = s4 * s3;
```

Finally, at the end of the trace, all values represented by the symbols stack are pushed to the operand stack. In this case, only `s5` is present on the symbol stack, and thus:

```
frame.opStack.push(s5);
```

is emitted. Putting it all together, we get:

```
var s1 = frame.opStack.pop();
var s2 = frame.opStack.pop();
var s3 = s1 + s2;
var s4 = frame.opStack.pop();
var s5 = s4 * s3;
frame.opStack.push(s5);
```

Even in this small example we can see that one opStack push has been eliminated for the `IADD` instruction. The savings are larger for larger traces.

## Benchmarks

I have used the following benchmarks:

* Compilation of a simple Java program using `javac` in `Chromium` and `Firefox`
* Scimark 2 on `nodejs`
* Bubble sort and back-tracking N-Queens on `nodejs`
* Running a Scala program on nodejs

## Benchmark results

| javac   | Total time for JVM | Total time for JVM (after) |
| -----------   | ------------------ | -------------------------- |
| Chromium 49   | 31994              | 17534                      |
| Firefox 45.0.1| 38283              | 24582                      |


| Scimark2      | Score before      | Score after        |
| -----------   | ----------------  | ------------------ |
| node 5.7.1    | 2.194548131157776 | 23.46              |

| Nashorn       | Time before   | Time after     |
| -----------   | ------------  | -------------- |
| node 5.7.1    | 24.85 s       | 18.645s        |

| Abandon.jar  | Time before   | Time after     |
| ----------------- | ------------  | -------------- |
| node 5.7.1        | 90.53s        | 71.22s         |

| Bubble sort + nQueens | Time before   | Total time for JVM | Time after     | Total time for JVM (after) |
| --------------------- | ------------  | ------------------ | -------------- | -------------------------- |
| node 5.7.1            | 56011 ms      | 57.2 s             | 8248 ms        | 10.47 s                    |
### Future ideas

* Detection and optimisations of loops, if-else conditions, etc.
* Inlining functions calls.

----

The source code and binaries used for benchmarking, along with a README, are available [here](https://hrj.github.io/static/benchmark-files.tar.bz2).

jvilk commented:
As mentioned in the previous closed PR, I'm planning to dive into this later on this week. I can't wait to test it out!

jvilk commented:
I've started looking at this now! My first comments are purely organizational in nature; I have not thought deeply about the JIT mechanisms yet. (That is coming next!)

I am slightly worried about bifurcating opcode logic into multiple files for maintenance reasons. I wonder if it would be cleaner to have a `class` for each opcode type with `static`-only methods and fields? That would collect the following information for each opcode into one place:

* The interpreter function
* The function that emits JavaScript for the JIT.
* Useful JIT properties (e.g. whether or not it is a JUMP instruction)
  * Then, you could change the code in the JIT that hardcodes particular opcodes to simply check if the opcode has a certain property (e.g. `if (opcodetype.jumps())` instead of `if (opcode === INVOKESTATIC || (etc))`
(etc.)

Each `class` itself would implement a consistent interface, so the result would be strongly typed as well.

The interpreter would gather all of the static interpreter functions into its lookup table like it does now, so it would not change any of that logic; it would just change the source of the interpreter functions.

Do you think that would result in cleaner, easier-to-maintain code?

jvilk commented:
I'm going to make this a meta-comment with a checklist, to be periodically updated as I notice things.

* [ ] (If agreed) Consolidate opcode logic into `class`es with `static` methods and fields.
* [ ] Expose `-Xint` option to switch off the JIT and confine execution to the interpreter.

jvilk commented on src/methods.ts:
```diff
@@ -226,6 +227,98 @@ export class Field extends AbstractMethodField {
+
+      // console.log(`Emitted trace of ${this.infos.length} ops: ` + emitted);
+      // f = frame, t = thread, u = util
+      return new Function("f", "t", "u", emitted);
```
`new Function()` is nice because it does not have the same scoping issues that `eval` has, and is much more friendly for JavaScript JITs.

The problem with `new Function()` is that there's no way to add a line to the end of the body to force Chrome DevTools to surface the code as a particular document name, making it hard to debug. If you open Chrome DevTools with Doppio currently, you'll notice that classes and other eval'd code have a friendly name in the devtools.

I'm fine with keeping `Function`, but I wanted to note that in case you were wondering why I used `eval` elsewhere.

hrj replied:
Indeed, `eval` has better debugging support. I chose `new Function()` for its performance, after a lot of benchmarking. You can see the results and benchmarking code [here](http://git.javadeploy.net/hrjet9/doppio-jit/issues/1). (It is not super descriptive; it was mainly intended for my own reference. To give a brief summary: I implemented a mini interpreter and various versions of a JITed function to compare their speed across Chromium, Firefox and node).

For debugging, I agree, `new Function()` has poor support. Though, Chromium does assign it a name automatically, with clickable links, if there is an exception. And Firefox has a non-standard feature called [displayName](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/displayName) that helps.

jvilk commented on src/threading.ts:
```diff
@@ -247,11 +270,18 @@ export class BytecodeStackFrame implements IStackFrame {
-      var op = code.readUInt8(this.pc);
-      if (!RELEASE && logging.log_level === logging.VTRACE) {
-        vtrace(`  ${this.pc} ${annotateOpcode(op, this, code, this.pc)}`);
+      var op = method.getOp(this.pc, code);
```
Counting # of opcodes executed seems excessive. I would be more in favor of number of function executions or, if you want to handle long-running loops, basic block executions.

hrj replied:
Function executions will not account for long-running loops as you rightly said.

Counting basic block executions is not possible currently, because the basic blocks (traces) are being created lazily after reaching a threshold. And the number of opcode executions is the threshold :)

It is possible to eagerly create basic blocks for all methods, but that would introduce a lot of overhead; most of the methods executed don't run long nor frequently enough.

hrj replied:
I just had another idea: count function executions + jumps with negative offsets. It is simple enough to not have an overhead, and specific enough to improve JIT targets.

Will try it out and push to this PR if it improves performance.

jvilk replied:
For basic blocks, you can just modify the interpreter instructions for jump opcodes with a line that increments the counter during jumps. That's how most JITs do it.

hrj replied:
Oh I see; I haven't seen any JITs before. I have coded the changes for counter-increments-during-jumps and will push tomorrow. Although I don't see any significant performance improvements, we might be able to tune it further post-merge.

jvilk commented on src/methods.ts:
```diff
@@ -254,6 +347,11 @@ export class Method extends AbstractMethodField {
 
+  private numOpcodeExecs = 0;
+  private jitThreshold = 0;
+  private compiledFunctions: Function[] = [];
```
If you made this a map from `startopcode` to `string`s with the implementation of that basic block, your JIT could produce a single `run` function that does something like:

```{js}
switch (pc) {
  case 10:
    ${this.compiledBlocks[10]}
    break;
  case 20:
    ${this.compiledBlocks[20]}
    break;
  default:
    this.interpret();
}
```

hrj replied:
I am not sure I understood this. Are you suggesting that we cache the JIT emitted function's source, instead of caching the compiled function?

hrj replied:
Although I don't understand the benefit of caching the function's source, I agree it would be nice to have a JIT specific run function with the `switch` statements. I had tried a JIT specific run function without `switch` and that had not improved any performance.

Will try the "run with switch" idea and push to this PR if it makes a difference.

jvilk replied:
I've actually experimented with a DoppioJVM JIT before. Caching the basic block source lets you progressively JIT more of the function, if you so choose. For example, if a basic block has not run, it still has `invokestatic` and friends instead of `invokestatic_fast`. You may want to re-JIT those basic blocks if they suddenly become hot, since the conversion happens on their first run.

If you are not doing that, then there's no reason to store the source.

hrj replied:
I do have a branch that re-JITs whenever byte code is updated to one of the "fast" opcodes. However, I didn't observe significant performance improvement with it.

My version of re-JIT emits the function source again. Using the previous source is not directly possible here because of lost information about the symbol-stack. We would need to keep that state information but for the first cut I thought it would be too complex.

jvilk replied:
> My version of re-JIT emits the function source again.

Gotcha; that's fine with me. It reduces memory usage, too.

> I do have a branch that re-JITs whenever byte code is updated to one of the "fast" opcodes. However, I didn't observe significant performance improvement with it.

In this case, it's all up to the heuristic you are using. If the method *interprets* a significant number of basic blocks after JITting, then it is a good candidate for re-JITting.

One possible issue here is that if there are particular blocks that cannot be JITted due to blacklisted opcodes (e.g. `wide`). A hot function with one of these blacklisted opcodes may repeatedly try to be JITted and fail. In this case, we can have a `blacklist` boolean or something to prevent that from happening.

How does your `switch` statement handle jumps from interpreted to compiled basic blocks? I'm thinking we can make the `switch` smarter and have a case statement for each interpreted basic block; it could do something like:

```{js}
while (pc < endOfBB && pc >= startOfBB) {
  // interpret
}
```

Also, note that, unlike traditional basic blocks, DoppioJVM basic blocks end when there is a method call or a classloading event, since those are asynchronous. I believe classloading events are completely ruled out if the blocks you are compiling have been run once and have the `fast` versions of opcodes. It looks like you already use that same definition, since you close traces at function call boundaries.

jvilk commented on src/jvm.ts:
```diff
@@ -427,6 +427,11 @@ class JVM {
         return false;
       case JVMStatus.TERMINATING:
+
+        if (!RELEASE) {
```
You should add a command line flag for dumping these stats (e.g. `-XdumpJitStats` or something).

jvilk commented:
OK, I've looked this over. Overall, this is great work, and I am excited to tinker with it once it is merged! :)

Current blockers to merging:

* The heuristic used to determine if something should be JITed is expensive and should be cheaper
    * Using some combination of number of executions of the function/basic block and function size seems like a good idea. Maybe we should conduct a literature search for common JIT compiler heuristics to experiment with.
* There should be an API and command-line flag to disable JITting (`-Xint`)
* The JIT should produce a single function that supplants the method's `run` function to avoid the shenanigans it currently does in `getOp`. :)

Do these changes sound reasonable?

hrj commented:
> Do you think that would result in cleaner, easier-to-maintain code?

I am not sure.. in some ways it is nicer to have separate files for interpreted v/s JIT code. Atleast while developing, I found it easier to have separate files, because each file had its own pattern and searching by opcode was easy.

hrj commented:
> Useful JIT properties (e.g. whether or not it is a JUMP instruction)
> 
>     Then, you could change the code in the JIT that hardcodes particular opcodes to simply check if the opcode has a certain property (e.g. if (opcodetype.jumps()) instead of if (opcode === INVOKESTATIC || (etc)) (EDIT: I guess you have .hasBranches! But it would formalize all of that into one place.) (etc.)

This is already in place. There is an interface called `JitInfo` defined in `jit.ts` that captures this info.

The `INVOKE_*` methods are special cases because their information is more dynamic than the other opcodes. Specifically, the number of operands popped by an `invoke` opcode depends on the method being called, unlike most other opcodes which have a fixed number of "pops".

jvilk commented:
> This is already in place. There is an interface called JitInfo defined in jit.ts that captures this info.

I guess my frustration is that it's not easy to extend the current setup with other, more dynamic info without further hacks. For example, for opcodes that make a NULL check, expose a function that determines the symbolic value that it checks. Then, under certain conditions, you can remove redundant NULL checks.

Anyway, any such changes could come post-merge, if it makes sense. I'll remove it from the blocking list.

hrj commented:
> I guess my frustration is that it's not easy to extend the current setup with other, more dynamic info without further hacks. For example, for opcodes that make a NULL check, expose a function that determines the symbolic value that it checks. Then, under certain conditions, you can remove redundant NULL checks.

For that specific example, is it possible that javascript VM can figure it out and eliminate the check? I have had only a brief glance at the debug log of V8 and it does seem like they have some aggressive optimisations.

> 
> Anyway, any such changes could come post-merge, if it makes sense. I'll remove it from the blocking list.

Yes, I think we could do it post-merge. If the approach of this PR seems good, we could use it as a baseline for performance and incrementally add many more features to the JIT.

jvilk commented:
> is it possible that javascript VM can figure it out and eliminate the check?

I have my doubts, since the return value from the `isNull` method is used in a conditional. But there is a way to check this -- you can actually use [IR viewers for V8](http://mrale.ph/irhydra/1/) to take a look at the resulting IR to see if it removes branches.

I'm pretty sure it will not remove it, though.

>Yes, I think we could do it post-merge. If the approach of this PR seems good, we could use it as a baseline for performance and incrementally add many more features to the JIT.

Sounds good to me. I updated the checklist; once those items are addressed, I will merge this in!

jvilk commented on src/threading.ts:
```diff
@@ -245,15 +269,36 @@ export class BytecodeStackFrame implements IStackFrame {
+    } else {
+      // Run until we get the signal to return to the thread loop.
+      while (!this.returnToThreadLoop) {
+        var op = method.getOp(this.pc, code);
```
Any reason why your JIT still uses this `getOp` scheme with an array of functions instead of making a single function with a `switch`?

hrj replied:
> Any reason why your JIT still uses this getOp scheme with an array of functions instead of making a single function with a switch?

I was going to try coding it yesterday but ran out of hours :) Benchmarking and tuning the basic-block-counting changes took most of my time.

I will give it a shot today.

jvilk commented:
I'd like to get rid of those conditionals somehow (`if (offset < 0)`), but it's OK for now. :)

> I was not able to add -XX:+PrintCompilation option. I have committed my attempt in d0ea20a but the option parser is complaining about something I don't understand. I could do this post-merge.

Ahh, OK. I wrote that parser. Could you add it as `-XPrintCompilation` for now, and I'll fix the parser post-merge?

Thanks for being patient! We're so close to merging.

hrj commented:
> Could you add it as -XPrintCompilation for now, and I'll fix the parser post-merge?

I have started working on this, but I worry that this feature might affect performance adversely, so I wanted to clarify. The flag is stored in the `JVM`, which is accessible via the thread object. So to access the flag in the JIT compiler methods, we need to pass the `thread` object along the call chain. Is that correct? Or is there a better way to access such flags from a `Method`'s context?

> 
> Thanks for being patient! We're so close to merging.

No worries; this PR would have a big impact so it's good to take our time with it.

hrj commented:
Also, sorry I haven't been able to address all points raised by you. There are many ideas and options to explore here and the need to benchmark and discuss alternatives. It might be best to create separate issues for each topic after this gets merged.

jvilk commented:
No worries! I just wanted to make the merge blockers clear. I was not sure if you had forgotten when you commented describing your changes. I know I can be all over the map when examining something; I tend to comment to document my thoughts more than to pile more work onto you.

> I have started working on this, but I worry that this feature might affect performance adversely, so I wanted to clarify. The flag is stored in the JVM, which is accessible via the thread object. So to access the flag in the JIT compiler methods, we need to pass the thread object along the call chain. Is that correct? Or is there a better way to access such flags from a Method's context?

I hadn't thought about this detail!

Ideally, your JIT will be a class that the JVM passes options to, which it uses to enable/disable JIT options. That will be needed in the long run to facilitate various experimental JIT flags/options. Also, implementing the JIT as a class with a well-defined interface would make it possible to run the JIT in a WebWorker if we switch to asynchronous compilation -- although there are fun trade-offs there (e.g. if compilation is sufficiently fast, switching to an async model will JIT functions slower and result in overall slower runtime). :)

For now, I do not think it would hurt to pass the `thread` to the compilation function, and have the `PrintCompilation` check guarded with an `if (!RELEASE)` to prevent slowdown in release builds.

hrj commented:
Update:
* I have implemented the -XPrintCompilation option and pushed the commit here
* I have implemented the `switch` loop based interpreter in 4186ddbd86e53bd5dd3e1856c79cf32c85cb64fe, but the performance degrades terribly. I guess the performance gain of using `switch` is offset by the time required to re-compile the interpreter function. It might be that my implementation is not right. Feel free to use that commit as a base; hopefully after merging this PR :)

All the blockers are done from my side!

jvilk commented:
Excellent! Now to make it easy to run benchmarks...
